File: 2025/02/p2/main.py
#!/bin/python3
# https://adventofcode.com/2025/day/2#part2

from os import path
from collections.abc import Iterator


DIVISORS: dict[int, list[int]] = {
        2: [1],
        3: [1],
        4: [1, 2],
        5: [1],
        6: [1, 2, 3]
    }


def get_divisors(idLen: int) -> list[int]:
    if idLen < 2:
        return []
    if idLen not in DIVISORS.keys():
        DIVISORS[idLen] = [1]
        for i in range(2, int(idLen/2) + 1):
            if idLen % i != 0:
                continue
            DIVISORS[idLen].append(i)
    return DIVISORS[idLen]


def get_ranges(file_name: str) -> Iterator[tuple[int, int]]:
    data: str = ''
    with open(path.join('..', 'data', file_name), mode='r', encoding='utf-8') as file:
        data = file.read().strip()

    start: int = 0
    stop: int = 0
    separator: int = 0
    while True:
        i = data.find(',', start)
        stop = i if i >= 0 else len(data)
        separator = data.find('-', start, stop)
        assert separator > 0, "no separator found!"
        yield (int(data[start:separator]), int(data[separator + 1:stop]))
        if i < 0:
            return
        start = stop + 1


def get_invalid_ids(start: int, stop: int) -> Iterator[int]:
    for number in range(start, stop + 1):
        # print(f"\t{number=}")
        s = str(number)
        for divisor in get_divisors(len(s)):
            parts = list(map(''.join, zip(*[iter(s)]*divisor)))
            isValid = len(set(parts)) > 1
            if number == 99998:
                print(f"\t\t{divisor=}, {isValid=}, {parts=}")
            if not isValid:
                yield number
                break


def main(file_name: str) -> None:
    s: int = 0
    for (start, stop) in get_ranges(file_name):
        print(f"{start}-{stop} ")
        for m in get_invalid_ids(start, stop):
            print(f"\t\t{m=}")
            s += m
    print(f"{s=}")


if __name__ == '__main__':
    # print(get_divisors(len(str(262626))))
    # main('example.txt')
    main('input.txt')

File: 2025/02/p2/rs/src/main.rs
use std::collections::HashMap;
use std::fs::File;
use std::io::{BufReader, BufRead};
use std::path::PathBuf;

fn open_file() -> File {
    let args: Vec<String> = std::env::args().collect();
    let file_name: &str;
    if args.len() <= 1 || args[1].len() == 0 {
        file_name = "example.txt"
    } else {
        file_name = args[1].as_str();
    }
    let path: PathBuf = ["..", "..", "..", "data", file_name].iter().collect();

    println!("opening {:?}", path.canonicalize());
    let file: File = match File::open(&path) {
        Err(e) => panic!("could not open {:?}: {}", path.canonicalize(), e),
        Ok(file) => file,
    };

    return file;
}

fn get_data() -> Vec<(u64, u64)> {
    let file = open_file();
    let reader = BufReader::new(file);
    let mut line_no = 0;

    let mut ret: Vec<(u64, u64)> = Vec::new();
    for line in reader.lines() {
        line_no += 1;

        let content: String = match line {
            Err(_) => panic!("could not read line no={}", line_no),
            Ok(content) => content,
        };

        let s: usize = match content.find("-") {
            None => panic!("no separator found in line no {}", line_no),
            Some(pos) =>  match pos {
                0 => panic!("invalid separator placement in line no {}", line_no),
                pos => pos,
            },
        };

        let start: u64 = match content[..s].parse::<u64>() {
            Err(_) => panic!("could not parse line no {}", line_no),
            Ok(value) => value,
        };

        let stop: u64 = match content[s+1..].parse::<u64>() {
            Err(_) => panic!("could not parse line no {}", line_no),
            Ok(value) => value,
        };

        ret.push((start, stop));
    }

    return ret;
}

fn get_divisors(length: usize, cache: &mut HashMap<usize, Vec<usize>>) -> &Vec<usize> {
    if !&cache.contains_key(&length) {
        let mut divs = Vec::<usize>::from([1]);
        for i in 2..(length / 2 + 1) {
            if length % i == 0 {
                divs.push(i);
            }
        }
        cache.insert(length, divs);
    }

    if let Some(ret) = &cache.get(&length) {
        return ret;
    }

    panic!("could not obtain list of divisors");
}

fn main() {
    let mut score: u64 = 0;
    let mut divisors_cache = HashMap::<usize, Vec<usize>>::from([
        (1, Vec::new()),
    ]);

    for (start, stop) in get_data() {
        for n in start..=stop {
            let s: String = n.to_string();
            for d in get_divisors(s.len(), &mut divisors_cache) {
                let chunks = s.chars()
                              .collect::<Vec<char>>()
                              .chunks(*d)
                              .map(|c| c.iter().collect::<String>())
                              .collect::<Vec<String>>();

                if chunks[1..].into_iter().all(|e| *e == chunks[0]) {
                    score += n as u64;
                    break;
                }
            }
        }
    }

    println!("score={}", score);
}
